staging/src/k8s.io/apiserver/pkg/server/filters/goaway_test.go

```diff
@@ -195,55 +345,26 @@ func TestClientReceivedGOAWAY(t *testing.T) {
 			// localAddr indicates how many TCP connection set up
 			localAddr := make([]string, 0)
 
-			// init HTTP2 client
-			client := http.Client{
-				Transport: &http2.Transport{
-					TLSClientConfig: tlsConfig,
-					DialTLS: func(network, addr string, cfg *tls.Config) (conn net.Conn, err error) {
-						conn, err = tls.Dial(network, addr, cfg)
-						if err != nil {
-							t.Fatalf("unexpect connection err: %v", err)
-						}
-						localAddr = append(localAddr, conn.LocalAddr().String())
-						return
-					},
-				},
+			client, err := newGOAWAYClient(func(network, addr string, cfg *tls.Config) (conn net.Conn, err error) {
+				conn, err = tls.Dial(network, addr, cfg)
+				if err != nil {
+					t.Fatalf("unexpect connection err: %v", err)
+				}
+				localAddr = append(localAddr, conn.LocalAddr().String())
```

Comment:
this added a data race on unguarded localAddr, fixed in https://github.com/kubernetes/kubernetes/pull/94530

Reply:
As far as I know, it's not the root cause of TestClientReceivedGOAWAY flake, because dial a new connection only when client send a new request, and in this case, requests are sent one by one.

I'm still debug the root cause of TestClientReceivedGOAWAY flake, for now, I doubt that I used an incorrect http2 server/client configuration(`http.Transport.MaxConnsPerHost=1` or SETTINGS_MAX_CONCURRENT_STREAMS limited), or the dial function ignored context.Context which will cause more than one goroutine dial the connection when the first canceled.

Another interesting thing is I got a data race result that write at `if tc.expectConnections != len(localAddr)`(main testing goroutine)  and read at `localAddr = append(localAddr, conn.LocalAddr().String())`, how does client try to set-up a new connection when we finished all requests?